### PR TITLE
feat: chroot isolation

### LINKED ISSUE #2165: Conflicting files while building images with Kanino on Kubernetes
**Actual behavior**
I'm using Tekton+EKS so we use kaniko to build images. 

We started to see an error like this:
```
{"level":"info","msg":"Running: [/bin/sh -c mkdir /root/.ssh]","time":"2022-07-06T14:22:19Z"}
mkdir: can't create directory '/root/.ssh': File exists
```

I checked the base image, and it does *not* have this folder.
Adding a `RUN ls -lah /root && ls -lah /root/.ssh` before this `RUN mkdir /root/.ssh` step, I got:
```
{"level":"info","msg":"Running: [/bin/sh -c ls -lah /root \u0026\u0026 ls -lah /root/.ssh ||:]","time":"2022-07-06T14:22:19Z"}
total 0      
drwx------    3 root     root          18 Jul  6 14:22 .
drwxr-xr-x    1 root     root         272 Jul  6 14:22 ..
drwx------    2 root     root          72 Jul  6 14:22 .ssh
total 12K    
drwx------    2 root     root          72 Jul  6 14:22 .
drwx------    3 root     root          18 Jul  6 14:22 ..
-rw-------    1 root     root         114 Jul  6 14:22 config
-rw-------    1 root     root        3.2K Jul  6 14:22 id_default-github-ssh-key
-rw-------    1 root     root         442 Jul  6 14:22 known_hosts
```

But these files are injected via ServiceAccount into kaniko's step, so it doesn't come from the base image. I tested changing to `mkdir -p /root/.ssh` and it passed, and it didn't add those files (thankfully)

**Expected behavior**
I expected some sort of isolation between what's run for building the layer. So I expected to be able to run `RUN mkdir <path>` if that file path doesn't exist on the base image.

**To Reproduce**
Although I'm using a Kubernetes, Tekton, etc. I believe you can safely test it locally by going to an empty folder and running this script:
```bash
#!/usr/bin/env bash
set -euo pipefail

mkdir -p ./foo

echo -e 'FROM alpine \nRUN mkdir /foo' > Dockerfile | tar -cf - Dockerfile | gzip -9 | docker run \
	--interactive -v $(pwd):/workspace -v $(pwd)/foo:/foo gcr.io/kaniko-project/executor:latest \
	--context tar://stdin \
	--log-format=json \
	--no-push \
	--tarPath=foo.tar \
	--destination=foo
```

 **Triage Notes for the Maintainers**
 
 | **Description** | **Yes/No** |
 |----------------|---------------|
 | Please check if this a new feature you are proposing        | <ul><li>- [ ] </li></ul>|
 | Please check if the build works in docker but not in kaniko | <ul><li>- [x] </li></ul>| 
 | Please check if this error is seen when you use `--cache` flag | <ul><li>- [ ] </li></ul>|
 | Please check if your dockerfile is a multistage dockerfile | <ul><li>- [ ] </li></ul>|

### PR DESCRIPTION
**Pull request by hown3d**
This PR implements isolation between build commands with chroot.
Should fix #2165 and #2153.

**Description**

Chroot isolation is implemented by creating a temporary directory inside the HOME directory which is later used as the base for chroot. Before chroot, the context directory, /proc and /dev are bind mounted into the new root directory.
Isolation is wrapped between each build command execution.

Isolation can be toggled with a flag **--isolation** which resolves by default to chroot. Providing anything else will disable isolation.

TODOs before removing Draft:
- [ ] Integration tests are passing
- [ ] New unit tests for chroot

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [ ] Adds integration tests if needed.

_See [the contribution guide](../CONTRIBUTING.md) for more details._

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit tests and or integration tests added.

**Release Notes**

User facing changes:
- kaniko adds a new flag `--isolation` to specify the isolation method. The default will be **chroot**.  Removing isolation can be done by setting the flag to **none**

### CODE REVIEW

File: pkg/chroot/root.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package chroot
+
+import (
+	"errors"
+	"os"
+)
+
+func TmpDirInHome() (string, error) {
```
Comment (imjasonh) [inline pkg/chroot/root.go:8]: We may want [`os.UserHomeDir()`](https://pkg.go.dev/os#UserHomeDir) instead, which uses `$HOME` and other heuristics.

File: pkg/executor/build.go [hunk 2]
```diff
@@ -348,94 +369,137 @@ func (s *stageBuilder) build() error {
...
-		isCacheCommand := func() bool {
-			switch command.(type) {
-			case commands.Cached:
-				return true
-			default:
-				return false
+func (s *stageBuilder) isolate() (exitFunc func() error, err error) {
+	switch s.opts.Isolation {
+	case "chroot":
+		{
```
Comment (imjasonh) [inline pkg/executor/build.go:395]: These `{}`s are unnecessary:

```
switch s.opts.Isolation {
case "chroot":
  newRoot, err := ...
  ...
case "none":
  ...
}
```

**Review by imjasonh — COMMENTED** [review]
This looks promising! Thanks for working on this. 👍

**Comment by imjasonh** [comment]
> During my development on this PR, I ran more and more into duplications with established isolation tools like runc and crun. Is it maybe worth considering investing into using those tools that are conform with the OCI spec standard? Reimplementing those isolation techniques is imo not easy for the long run.
> 
> I'll look into the possibilities with runc's libcontainer and how it plays with kaniko.

First of all, thanks for digging into this!

I'm excited to learn what our options are if we want to retain the invariant that Kaniko doesn't require the kind of privilege that `docker build`, Buildah, etc., require, and expects to be run inside a container itself. This is an area I'm not very experienced with, so I'll mostly defer to your expertise and findings here.

**Comment by hown3d** [comment]
> > During my development on this PR, I ran more and more into duplications with established isolation tools like runc and crun. Is it maybe worth considering investing into using those tools that are conform with the OCI spec standard? Reimplementing those isolation techniques is imo not easy for the long run.
> > I'll look into the possibilities with runc's libcontainer and how it plays with kaniko.
> 
> First of all, thanks for digging into this!
> 
> I'm excited to learn what our options are if we want to retain the invariant that Kaniko doesn't require the kind of privilege that `docker build`, Buildah, etc., require, and expects to be run inside a container itself. This is an area I'm not very experienced with, so I'll mostly defer to your expertise and findings here.

I took a look around for other isolation tools (libcontainer, bubblewrap, crun) but all of those tools need privileged containers, because they create cgroups and those need writable /sys dirs, thus --privileged.

I'll try to implement a sort of style that buildah did with their [chroot isolation](https://github.com/containers/buildah/tree/main/chroot).

Integration tests for isolation will be created aswell (breaking out of chroot, accessing "host" files).

**Comment by natalieparellano** [comment]
I apologize for my ignorance here, but would it be possible to explain at a high level how does this PR circumvent the limitations noted here: https://github.com/GoogleContainerTools/kaniko/issues/107#issuecomment-384427886? Would a platform running containers without `--privileged` need to pass `--isolation none`?

**Comment by hown3d** [comment]
Glad you asked, since this is a Draft PR I'm not 100% settled on the final implementation, but the current idea is the following:
1. Unpack Rootfs of image into isolated directory under home directory of user 
2. Create user & mount namespace to map current user to root (future rootless implementation?) and bind mount /proc /sys /dev /etc/resolv.conf and /etc/hosts (networking purposes) into the new root
3. pivot_root into newly created isolated directory where the rootfs lays and set the capabilities to the same ones that docker uses (because unshare creates a full set of capabilities)

Currently, this implementation has to set the seccomp and apparmor profiles to "unconfined" because those profiles are very strict in docker (CRI-O, Podman etc. allow mount syscalls), but it's being discussed in the moby repo: https://github.com/moby/moby/pull/42455

The final goal is to be able to run --isolation chroot without SYS_ADMIN privilege and with a slightly different seccomp/apparmor profile than default (when running with docker).

Also I think it's best to leave the default to --isolation **none** at first, since changing the default could break builds.

**Comment by gabyx** [comment]
Cool stuff: Some noobie question: Why is chroot isolation needed, what does it achieve? I know it solves a security issue...

**Comment by Lord-Y** [comment]
Hello guys, any news about this PR?

**Comment by gewei2000** [comment]
hi guys, is there any progress about this PR?

**Comment by hown3d** [comment]
Hey everyone,
I haven't had the time working on that PR, but I'm also reconsidering if this is still a good idea to implement.
The thing that bothers me:
- Chroot itself is not enough of a isolation
-> escaping from chroot is easily done when running as root, so we need to run kaniko as a non root user.
When running as non-root we would need to implement running in user namespaces to achieve root-like behavior.
This leads us into container in container issues, like having to run with a customized seccomp profile, because docker default seccomp profile blocks *setns* and *unshare* syscalls.

Kaniko is just not designed to provide more isolation inside the container environment. 
If you want to build dockerfiles inside containers in a rootless way with more isolation inside the container, but more privileges on the host, consider using the following options:
- [buildah](https://github.com/containers/buildah/blob/main/docs/buildah-build.1.md) rootless with chroot isolation (needs customized seccomp and selinux disabled).
The _containers_ team provides a container image with instructions on [quay.io](https://quay.io/repository/containers/buildah) 
- [buildkit rootless in container](https://github.com/moby/buildkit#containerizing-buildkit)
Can be configured nearly in the same way as the buildah instructions aboth. Uses [rootlesskit](https://github.com/rootless-containers/rootlesskit) to enter new namespaces inside container

I'm hereby closing this PR, since I won't continue working on that feature.

**Pull request closed without merging**